### ISSUE TITLE
[pytorch upstream] softmax bwd on BMG is slow

### ISSUE THREAD

**Issue by jianyizh** [intel/intel-xpu-backend-for-triton#5275]
### Describe the issue

Similar to #5133, the backward of slowmax is also slow on BMG when run cait_m36_384 training on BMG compared to 4080 super. (0.354ms vs 0.176ms, ~49%), according to peak bandwidth of two hardware, we should at least reach ~57%

reproduce on XPU:
```python
import triton
import triton.language as tl

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, DeviceProperties
triton_helpers.set_driver_to_gpu()

from torch._dynamo.testing import rand_strided
from torch._C import _xpu_getCurrentRawStream as get_raw_stream
import torch

@triton_heuristics.reduction(
    size_hints={'y': 16, 'x': 1024, 'r0_': 2048},
    reduction_hint=ReductionHint.OUTER,
    filename=__file__,
    triton_meta={'signature': {'in_ptr0': '*fp16', 'in_ptr1': '*fp32', 'in_ptr2': '*fp16', 'out_ptr0': '*fp32', 'ynumel': 'i32', 'xnumel': 'i32', 'r0_numel': 'i32', 'YBLOCK': 'constexpr', 'XBLOCK': 'constexpr', 'R0_BLOCK': 'constexpr'}, 'device': DeviceProperties(type='xpu', index=0, multi_processor_count=20, cc={'architecture': 21479031808, 'device_id': 57867, 'driver_version': '1.6.35096+9', 'gpu_eu_count': 160, 'gpu_subslice_count': 20, 'has_atomic64': True, 'has_bfloat16_conversions': True, 'has_fp16': True, 'has_fp64': True, 'has_subgroup_2d_block_io': True, 'has_subgroup_matrix_multiply_accumulate': True, 'has_subgroup_matrix_multiply_accumulate_tensor_float32': False, 'max_compute_units': 160, 'max_num_sub_groups': 64, 'max_work_group_size': 1024, 'name': 'Intel(R) Arc(TM) B580 Graphics', 'platform_name': 'Intel(R) oneAPI Unified Runtime over Level-Zero', 'sub_group_sizes': [16, 32], 'total_memory': 12168933376, 'type': 'gpu', 'vendor': 'Intel(R) Corporation', 'version': '20.1.0'}, major=None, regs_per_multiprocessor=None, max_threads_per_multi_processor=None, warp_size=32), 'constants': {}, 'configs': [{(0,): [['tt.divisibility', 16]], (1,): [['tt.divisibility', 16]], (2,): [['tt.divisibility', 16]], (3,): [['tt.divisibility', 16]], (4,): [['tt.divisibility', 16]], (6,): [['tt.divisibility', 16]]}]},
    inductor_meta={'grid_type': 'Grid2D', 'autotune_hints': set(), 'kernel_name': 'triton_red_fused__softmax_backward_data_permute_sum_view_42', 'mutated_arg_names': [], 'optimize_mem': True, 'no_x_dim': False, 'num_load': 3, 'num_reduction': 1, 'backend_hash': 'FBB42C955946FB90F6347682AF516A12CAD552ECECB73587A3F0491DF8B215AE', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': True, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'tiling_scores': {'y': 42467328, 'x': 82944, 'r0_': 42467328}, 'kernel_num_gb': 0.085123584, 'kernel_flop': 0}
)
@triton.jit
def triton_red_fused__softmax_backward_data_permute_sum_view_42(in_ptr0, in_ptr1, in_ptr2, out_ptr0, ynumel, xnumel, r0_numel, YBLOCK : tl.constexpr, XBLOCK : tl.constexpr, R0_BLOCK : tl.constexpr):
    ynumel = 16
    xnumel = 648
    r0_numel = 2048
    rnumel = r0_numel
    RBLOCK: tl.constexpr = R0_BLOCK
    yoffset = tl.program_id(1) * YBLOCK
    yindex = yoffset + tl.arange(0, YBLOCK)[:, None, None]
    ymask = yindex < ynumel
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[None, :, None]
    xmask = xindex < xnumel
    r0_base = tl.arange(0, R0_BLOCK)[None, None, :]
    rbase = r0_base
    x1 = xindex
    y0 = yindex
    _tmp10 = tl.full([YBLOCK, XBLOCK, R0_BLOCK], 0, tl.float32)
    for r0_offset in range(0, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_2 = r0_index
        tmp0 = tl.load(in_ptr0 + (331776*y0 + 5308416*((r0_2 + 2048*x1) // 331776) + (((r0_2 + 2048*x1) % 331776))), r0_mask & xmask & ymask, eviction_policy='evict_last', other=0.0).to(tl.float32)
        tmp3 = tl.load(in_ptr1 + (576*y0 + 9216*((r0_2 + 2048*x1) // 331776) + ((((r0_2 + 2048*x1) // 576) % 576))), r0_mask & xmask & ymask, eviction_policy='evict_last', other=0.0)
        tmp4 = tl.load(in_ptr2 + (y0 + 16*r0_2 + 32768*x1), r0_mask & xmask & ymask, eviction_policy='evict_last', other=0.0).to(tl.float32)
        tmp1 = tmp0.to(tl.float32)
        tmp2 = -tmp1
        tmp5 = tmp4.to(tl.float32)
        tmp6 = tmp5 * tmp1
        tmp7 = libdevice.fma(tmp2, tmp3, tmp6)
        tmp8 = tmp7.to(tl.float32)
        tmp9 = tl.broadcast_to(tmp8, [YBLOCK, XBLOCK, R0_BLOCK])
        tmp11 = _tmp10 + tmp9
        _tmp10 = tl.where(r0_mask & xmask & ymask, tmp11, _tmp10)
    tmp10 = tl.sum(_tmp10, 2)[:, :, None]
    tl.store(out_ptr0 + (x1 + 648*y0), tmp10, xmask & ymask)


def get_args():
    arg_0 = rand_strided((4, 16, 576, 576), (5308416, 331776, 576, 1), device='xpu:0', dtype=torch.float16)
    arg_1 = rand_strided((4, 16, 576, 1), (9216, 576, 1, 36864), device='xpu:0', dtype=torch.float32)
    arg_2 = rand_strided((1327104, 16), (16, 1), device='xpu:0', dtype=torch.float16)
    arg_3 = rand_strided((1, 1, 1, 16, 648), (10368, 10368, 10368, 648, 1), device='xpu:0', dtype=torch.float32)
    return arg_0, arg_1, arg_2, arg_3, 16, 648, 2048,


def call(args):
    with torch.xpu._DeviceGuard(0):
        torch.xpu.set_device(0)
        stream0 = get_raw_stream(0)
        triton_red_fused__softmax_backward_data_permute_sum_view_42.run(*args, stream=stream0)


def benchmark_all_configs(args):
    with torch.xpu._DeviceGuard(0):
        torch.xpu.set_device(0)
        return triton_red_fused__softmax_backward_data_permute_sum_view_42.benchmark_all_configs(*args)


if __name__ == '__main__':
    from torch._inductor.runtime.benchmarking import benchmarker

    args = get_args()
    ms = benchmarker.benchmark_gpu(lambda: call(args), rep=40)
    num_gb = 0.085123584
    gb_per_s = num_gb / (ms / 1e3)
    print(f"{ms:.3f}ms    {num_gb:.3f}GB    {gb_per_s:.2f}GB/s")

```

reproduce on cuda:
```python
import triton
import triton.language as tl

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, DeviceProperties
triton_helpers.set_driver_to_gpu()

from torch._dynamo.testing import rand_strided
from torch._C import _cuda_getCurrentRawStream as get_raw_stream
import torch

@triton_heuristics.reduction(
    size_hints={'y': 16, 'x': 1024, 'r0_': 2048},
    reduction_hint=ReductionHint.OUTER,
    filename=__file__,
    triton_meta={'signature': {'in_ptr0': '*fp16', 'in_ptr1': '*fp32', 'in_ptr2': '*fp16', 'out_ptr0': '*fp32', 'ynumel': 'i32', 'xnumel': 'i32', 'r0_numel': 'i32', 'YBLOCK': 'constexpr', 'XBLOCK': 'constexpr', 'R0_BLOCK': 'constexpr'}, 'device': DeviceProperties(type='cuda', index=0, multi_processor_count=80, cc=89, major=8, regs_per_multiprocessor=65536, max_threads_per_multi_processor=1536, warp_size=32), 'constants': {}, 'configs': [{(0,): [['tt.divisibility', 16]], (1,): [['tt.divisibility', 16]], (2,): [['tt.divisibility', 16]], (3,): [['tt.divisibility', 16]], (4,): [['tt.divisibility', 16]], (6,): [['tt.divisibility', 16]]}]},
    inductor_meta={'grid_type': 'Grid2D', 'autotune_hints': set(), 'kernel_name': 'triton_red_fused__softmax_backward_data_permute_sum_view_37', 'mutated_arg_names': [], 'optimize_mem': True, 'no_x_dim': False, 'num_load': 3, 'num_reduction': 1, 'backend_hash': '7606AC1BD735D3E3F140115999815ACFE642967D9047962703B386A670225BF4', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': True, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'tiling_scores': {'y': 42467328, 'x': 82944, 'r0_': 42467328}, 'kernel_num_gb': 0.085123584, 'kernel_flop': 0}
)
@triton.jit
def triton_red_fused__softmax_backward_data_permute_sum_view_37(in_ptr0, in_ptr1, in_ptr2, out_ptr0, ynumel, xnumel, r0_numel, YBLOCK : tl.constexpr, XBLOCK : tl.constexpr, R0_BLOCK : tl.constexpr):
    ynumel = 16
    xnumel = 648
    r0_numel = 2048
    rnumel = r0_numel
    RBLOCK: tl.constexpr = R0_BLOCK
    yoffset = tl.program_id(1) * YBLOCK
    yindex = yoffset + tl.arange(0, YBLOCK)[:, None, None]
    ymask = yindex < ynumel
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[None, :, None]
    xmask = xindex < xnumel
    r0_base = tl.arange(0, R0_BLOCK)[None, None, :]
    rbase = r0_base
    x1 = xindex
    y0 = yindex
    _tmp10 = tl.full([YBLOCK, XBLOCK, R0_BLOCK], 0, tl.float32)
    for r0_offset in range(0, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_2 = r0_index
        tmp0 = tl.load(in_ptr0 + (331776*y0 + 5308416*((r0_2 + 2048*x1) // 331776) + (((r0_2 + 2048*x1) % 331776))), r0_mask & xmask & ymask, eviction_policy='evict_last', other=0.0).to(tl.float32)
        tmp3 = tl.load(in_ptr1 + (576*y0 + 9216*((r0_2 + 2048*x1) // 331776) + ((((r0_2 + 2048*x1) // 576) % 576))), r0_mask & xmask & ymask, eviction_policy='evict_last', other=0.0)
        tmp4 = tl.load(in_ptr2 + (y0 + 16*r0_2 + 32768*x1), r0_mask & xmask & ymask, eviction_policy='evict_last', other=0.0).to(tl.float32)
        tmp1 = tmp0.to(tl.float32)
        tmp2 = -tmp1
        tmp5 = tmp4.to(tl.float32)
        tmp6 = tmp5 * tmp1
        tmp7 = libdevice.fma(tmp2, tmp3, tmp6)
        tmp8 = tmp7.to(tl.float32)
        tmp9 = tl.broadcast_to(tmp8, [YBLOCK, XBLOCK, R0_BLOCK])
        tmp11 = _tmp10 + tmp9
        _tmp10 = tl.where(r0_mask & xmask & ymask, tmp11, _tmp10)
    tmp10 = tl.sum(_tmp10, 2)[:, :, None]
    tl.store(out_ptr0 + (x1 + 648*y0), tmp10, xmask & ymask)


def get_args():
    arg_0 = rand_strided((4, 16, 576, 576), (5308416, 331776, 576, 1), device='cuda:0', dtype=torch.float16)
    arg_1 = rand_strided((4, 16, 576, 1), (9216, 576, 1, 36864), device='cuda:0', dtype=torch.float32)
    arg_2 = rand_strided((1327104, 16), (16, 1), device='cuda:0', dtype=torch.float16)
    arg_3 = rand_strided((1, 1, 1, 16, 648), (10368, 10368, 10368, 648, 1), device='cuda:0', dtype=torch.float32)
    return arg_0, arg_1, arg_2, arg_3, 16, 648, 2048,


def call(args):
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        stream0 = get_raw_stream(0)
        triton_red_fused__softmax_backward_data_permute_sum_view_37.run(*args, stream=stream0)


def benchmark_all_configs(args):
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        return triton_red_fused__softmax_backward_data_permute_sum_view_37.benchmark_all_configs(*args)


if __name__ == '__main__':
    from torch._inductor.runtime.benchmarking import benchmarker

    args = get_args()
    ms = benchmarker.benchmark_gpu(lambda: call(args), rep=40)
    num_gb = 0.085123584
    gb_per_s = num_gb / (ms / 1e3)
    print(f"{ms:.3f}ms    {num_gb:.3f}GB    {gb_per_s:.2f}GB/s")

```

### Environment details

pytorch-triton-xpu 3.5.0+git1b0418a9
i libigc-dev 2.18.5-1188\~25.04 amd64 Core development files for Intel(R) Graphics Compiler for OpenCL(TM)
ii libze-intel-gpu1 25.35.35096.9\-1~25.04\~ppa3 amd64 Intel(R) Graphics Compute Runtime for oneAPI Level Zero.

Intel(R) Arc(TM) B580 Graphics

**Comment by dev-tomek** [comment]
The issue is gone on the newer Triton versions.

B580
triton-3.5.0+git1b0418a9                29th August
0.377ms    0.085GB    225.93GB/s

triton-3.5.0+gitc437c951                16th october
0.222ms    0.085GB    383.64GB/s
~79%  (>57%) of what was reported for 4080

**Comment by jianyizh** [comment]
It's still 0.35ms on my BMG with c437c951. Can you provide more details like driver, oneapi.... By the way, if there's more details on the root cause and optimization, could you put it here?

**Comment by dev-tomek** [comment]
libze-intel-gpu1 25.35.35096.9-1 24.04 ppa3 amd64
libigc2  2.18.5-1188~24.04 amd64
torch             2.10.0a0+git7f4c3e7
triton            3.5.0+gitc437c951
0.222ms    0.085GB    383.73GB/s

Seems like the ubuntu version is the only difference, however I am not sure that causes this issue. I have not performed any additional optimizations. Running the script on the commit you provided and then switching to a newer version resulted in performance improvement. Given it seemed resolved on the newer commits I did not perform bisection for the root cause.

Did you try clearing `~/.triton/cache/ ` after recompiling triton on https://github.com/intel/intel-xpu-backend-for-triton/commit/c437c9514ccc42aede56d37cf544d4ba0bbb81e1 ?

**Comment by jianyizh** [comment]
Yes, I cleaned triton cache before benchmarking. Maybe you can mark it as resolved for now, we can test later when upgrade triton commit pin in pytorch.
@chuanqi129 When we do acceptance test on triton update, can we also verify related issues such as this one is solved?
